File: RDMPEG/RDMPEGPlayer/RDMPEGPlayerView/RDMPEGPlayerView.m
//
//  RDMPEGPlayerView.m
//  RDMPEG
//
//  Created by Serhii Alpieiev on 9/13/17.
//  Copyright © 2017 Readdle. All rights reserved.
//

#import "RDMPEGPlayerView.h"
#import "RDMPEGRenderView.h"
#import <Log4Cocoa/Log4Cocoa.h>



NS_ASSUME_NONNULL_BEGIN

@interface RDMPEGPlayerView ()

@property (nonatomic, strong) UILabel *subtitleLabel;
@property (nonatomic, strong, nullable) RDMPEGRenderView *renderView;
@property (nonatomic, strong, nullable) NSString *subtitle;

@end



@implementation RDMPEGPlayerView

#pragma mark - Overridden Class Methods

+ (L4Logger *)l4Logger {
    return [L4Logger loggerForName:@"rd.mediaplayer.RDMPEGPlayerView"];
}

#pragma mark - Lifecycle

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        self.subtitleLabel = [[UILabel alloc] init];
        self.subtitleLabel.numberOfLines = 0;
        self.subtitleLabel.textColor = [UIColor whiteColor];
        self.subtitleLabel.backgroundColor = [UIColor colorWithWhite:0.0 alpha:0.5];
        self.subtitleLabel.clipsToBounds = YES;
        self.subtitleLabel.layer.cornerRadius = 2.0;
        [self addSubview:self.subtitleLabel];
    }
    return self;
}

#pragma mark - Overridden

- (void)layoutSubviews {
    [super layoutSubviews];
    
    if (self.subtitleLabel.text.length == 0) {
        self.subtitleLabel.frame = CGRectZero;
    }
    else {
        const CGFloat horisontalSubtitleOffset = 10.0;
        const CGFloat verticalSubtitleOffset = 10.0;
        
        CGRect aspectFitVideoFrame = self.renderView.aspectFitVideoFrame;
        
        CGFloat subtitleMinX = CGRectGetMinX(aspectFitVideoFrame) + horisontalSubtitleOffset;
        CGFloat subtitleMaxY = CGRectGetMaxY(aspectFitVideoFrame) - verticalSubtitleOffset;
        CGFloat subtitleMaxWidth = CGRectGetWidth(aspectFitVideoFrame) - horisontalSubtitleOffset * 2.0;
        CGFloat subtitleMaxHeight = CGRectGetHeight(aspectFitVideoFrame) - verticalSubtitleOffset * 2.0;
        CGSize subtitleFitSize = CGSizeMake(subtitleMaxWidth, subtitleMaxHeight);
        
        CGSize subtitleSize = [self.subtitleLabel sizeThatFits:subtitleFitSize];
        
        CGRect subtitleFrame = CGRectMake(subtitleMinX + (subtitleMaxWidth - subtitleSize.width) / 2.0,
                                          subtitleMaxY - subtitleSize.height,
                                          subtitleSize.width,
                                          subtitleSize.height);
        
        self.subtitleLabel.frame = CGRectIntegral(subtitleFrame);
    }
}

#pragma mark - Public Accessors

- (void)setAspectFillMode:(BOOL)aspectFillMode {
    if (_aspectFillMode == aspectFillMode) {
        return;
    }
    
    _aspectFillMode = aspectFillMode;
    
    self.renderView.aspectFillMode = _aspectFillMode;
}

- (void)setRenderView:(nullable RDMPEGRenderView *)renderView {
    if (_renderView == renderView) {
        return;
    }
    
    if (_renderView) {
        if ([_renderView isDescendantOfView:self]) {
            [_renderView removeFromSuperview];
        }
        else {
            log4Assert(NO, @"Content view is not descendant of player view");
        }
    }
    
    _renderView = renderView;
    
    if (_renderView) {
        _renderView.frame = self.bounds;
        _renderView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
        _renderView.aspectFillMode = self.isAspectFillMode;
        [self addSubview:_renderView];
        
        [self bringSubviewToFront:self.subtitleLabel];
    }
}

- (CGRect)videoFrame {
    return self.renderView.videoFrame;
}

- (void)setSubtitle:(nullable NSString *)subtitle {
    self.subtitleLabel.text = subtitle;
    
    [self setNeedsLayout];
}

- (nullable NSString *)subtitle {
    return self.subtitleLabel.text;
}

@end

NS_ASSUME_NONNULL_END

File: RDMPEG/RDMPEGPlayer/RDMPEGPlayerView/RDMPEGPlayerView.swift
//
//  RDMPEGPlayerView.swift
//  RDMPEG
//
//  Created by Max Berezhnoy on 18/07/2024.
//  Copyright © 2024 Readdle. All rights reserved.
//

import UIKit
import Log4Cocoa

@objc
public class RDMPEGPlayerView: UIView {
    @objc public var videoFrame: CGRect {
        return renderView?.videoFrame ?? .zero
    }

    @objc public var isAspectFillMode: Bool = false {
        didSet {
            if isAspectFillMode != oldValue {
                renderView?.isAspectFillMode = isAspectFillMode
            }
        }
    }

    private let subtitleLabel: UILabel
    var renderView: RDMPEGRenderView? {
        didSet {
            if renderView !== oldValue {
                oldValue?.removeFromSuperview()

                if let renderView = renderView {
                    renderView.frame = bounds
                    renderView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
                    renderView.isAspectFillMode = isAspectFillMode
                    addSubview(renderView)
                    bringSubviewToFront(subtitleLabel)
                }
            }
        }
    }

    var subtitle: String? {
        get { return subtitleLabel.text }
        set {
            subtitleLabel.text = newValue
            setNeedsLayout()
        }
    }

    override init(frame: CGRect) {
        subtitleLabel = UILabel()
        subtitleLabel.numberOfLines = 0
        subtitleLabel.textColor = .white
        subtitleLabel.backgroundColor = UIColor(white: 0.0, alpha: 0.5)
        subtitleLabel.clipsToBounds = true
        subtitleLabel.layer.cornerRadius = 2.0

        super.init(frame: frame)

        addSubview(subtitleLabel)
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override public func layoutSubviews() {
        super.layoutSubviews()

        if subtitleLabel.text?.isEmpty ?? true {
            subtitleLabel.frame = .zero
        }
        else {
            let horizontalSubtitleOffset: CGFloat = 10.0
            let verticalSubtitleOffset: CGFloat = 10.0

            guard let aspectFitVideoFrame = renderView?.aspectFitVideoFrame else { return }

            let subtitleMinX = aspectFitVideoFrame.minX + horizontalSubtitleOffset
            let subtitleMaxY = aspectFitVideoFrame.maxY - verticalSubtitleOffset
            let subtitleMaxWidth = aspectFitVideoFrame.width - horizontalSubtitleOffset * 2.0
            let subtitleMaxHeight = aspectFitVideoFrame.height - verticalSubtitleOffset * 2.0
            let subtitleFitSize = CGSize(width: subtitleMaxWidth, height: subtitleMaxHeight)

            let subtitleSize = subtitleLabel.sizeThatFits(subtitleFitSize)

            let subtitleFrame = CGRect(
                x: subtitleMinX + (subtitleMaxWidth - subtitleSize.width) / 2.0,
                y: subtitleMaxY - subtitleSize.height,
                width: subtitleSize.width,
                height: subtitleSize.height
            )

            subtitleLabel.frame = subtitleFrame.integral
            subtitleLabel.isHidden = (bounds.width < 120.0)
        }
    }
}

extension RDMPEGPlayerView {
    override public class func l4Logger() -> L4Logger {
        return L4Logger(forName: "rd.mediaplayer.RDMPEGPlayerView")
    }
}
